README.md
phantom-elliptics
=================

Elliptics module for phantom web server


debian/changelog
phantom-elliptics (1.1.0-ell2.26) unstable; urgency=low

  * Updated for phantom-0.14.0~pre65 & elliptics-2.26

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Tue, 12 Aug 2014 21:01:46 +0400

phantom-elliptics (1.0.11-ell2.24) unstable; urgency=low

  * Depend on elliptics 2.24.7 and higher

 -- Evgeniy Polyakov <zbr@ioremap.net>  Thu, 30 May 2013 00:03:24 +0400

phantom-elliptics (1.0.10-ell2.24) unstable; urgency=low

  * Build with 2.24 elliptics

 -- Evgeniy Polyakov <zbr@ioremap.net>  Sat, 04 May 2013 23:42:30 +0400

phantom-elliptics (1.0.9) unstable; urgency=low

  * Update to elliptics-2.23

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Fri, 05 Apr 2013 19:57:21 +0400

phantom-elliptics (1.0.8) unstable; urgency=low

  * Added ability to set id for exec command

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Thu, 14 Mar 2013 17:42:18 +0400

phantom-elliptics (1.0.7) unstable; urgency=low

  * Added exec command

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Wed, 13 Mar 2013 21:12:31 +0400

phantom-elliptics (1.0.6) unstable; urgency=low

  * Fixed time_recv value

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Tue, 19 Feb 2013 16:16:25 +0400

phantom-elliptics (1.0.5) unstable; urgency=low

  * Added dnet_id support

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Tue, 05 Feb 2013 03:13:31 +0400

phantom-elliptics (1.0.4) unstable; urgency=low

  * Added several nodes support

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Tue, 05 Feb 2013 02:25:14 +0400

phantom-elliptics (1.0.3) unstable; urgency=low

  * Added log_level property

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Tue, 05 Feb 2013 00:48:32 +0400

phantom-elliptics (1.0.2) unstable; urgency=low

  * Added support for ipv4/ipv6

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Fri, 01 Feb 2013 17:07:57 +0400

phantom-elliptics (1.0.1) unstable; urgency=low

  * Make config more similiar to method_stream

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Thu, 31 Jan 2013 19:50:08 +0400

phantom-elliptics (1.0.0) unstable; urgency=low

  * Initial release

 -- Ruslan Nigmatullin <euroelessar@yandex.ru>  Wed, 30 Jan 2013 23:45:58 +0400


src/source.H
// This file is part of the phantom::io_benchmark module.
// Copyright (C) 2013, Ruslan Nigmatullin <euroelessar@yandex.ru>.
// Copyright (C) 2013, YANDEX LLC.
// This module may be distributed under the terms of the GNU LGPL 2.1.
// See the file ‘COPYING’ or ‘http://www.gnu.org/licenses/lgpl-2.1.html’.

#pragma once

#include <phantom/pd.H>

#include <pd/base/out.H>
#include <pd/base/time.H>

#include <vector>

#pragma GCC visibility push(default)

namespace phantom { namespace io_benchmark { namespace method_elliptics {

enum command_t {
	write_data,
	read_data,
	remove_data,
	exec_request
};

struct request_t {
	command_t command;

	uint64_t cflags;
	uint32_t ioflags;
	std::vector<int> groups;

	in_segment_t filename;
	in_segment_t id;

	in_segment_t request;
	in_segment_t tag;

	uint64_t offset;
	uint64_t size;
	in_segment_t data;
};

class elliptics_source_t {
public:
	virtual bool get_request(request_t &request) const = 0;
	virtual void init() = 0;
	virtual void run() = 0;
	virtual void stat_print() const = 0;
	virtual void fini() = 0;

protected:
	inline elliptics_source_t() throw() { }
	inline ~elliptics_source_t() throw() { }

private: // don't use
	void *operator new(size_t);
	void operator delete(void *);
};

}}} // namespace phantom::io_benchmark::method_elliptics

#pragma GCC visibility pop


src/method_elliptics.C
// This file is part of the phantom::io_benchmark module.
// Copyright (C) 2013, Ruslan Nigmatullin <euroelessar@yandex.ru>.
// Copyright (C) 2013, YANDEX LLC.
// This module may be distributed under the terms of the GNU LGPL 2.1.
// See the file ‘COPYING’ or ‘http://www.gnu.org/licenses/lgpl-2.1.html’.

#include "method_elliptics.H"
#include "source.H"

#include <phantom/module.H>
#include <phantom/io_benchmark/method_stream/logger.H>
#include <pd/bq/bq_util.H>
#include <pd/bq/bq_cond.H>
#include <pd/base/stat_items.H>
#include <pd/base/exception.H>

static int dnet_parse_numeric_id(char *value, unsigned char *id)
{
	unsigned char ch[5];
	unsigned int i, len = strlen(value);

	memset(id, 0, DNET_ID_SIZE);

	if (len/2 > DNET_ID_SIZE)
		len = DNET_ID_SIZE * 2;

	ch[0] = '0';
	ch[1] = 'x';
	ch[4] = '\0';
	for (i=0; i<len / 2; i++) {
		ch[2] = value[2*i + 0];
		ch[3] = value[2*i + 1];

		id[i] = (unsigned char)strtol((const char *)ch, NULL, 16);
	}

	if (len & 1) {
		ch[2] = value[2*i + 0];
		ch[3] = '0';

		id[i] = (unsigned char)strtol((const char *)ch, NULL, 16);
	}

	return 0;
}

namespace phantom {
namespace io_benchmark {

MODULE(io_benchmark_method_elliptics);

method_elliptics_t::config_t::config_t() throw() :
	port(0), family(0), logger_filename(STRING("/dev/null")), logger_level(STRING("error")),
	timeout(5), nodes_count(1), check_timeout(20), flags(0), io_thread_num(1),
	net_thread_num(1) {
}

void method_elliptics_t::config_t::check(const in_t::ptr_t &ptr) const {
	if (!nodes_count)
		config::error(ptr, "nodes_count must be positive");

	if (!port)
		config::error(ptr, "port is required");

	if (!family)
		config::error(ptr, "family is required");

	if (!source)
		config::error(ptr, "source is required");
}

void method_elliptics_t::loggers_t::commit(const in_segment_t &request, const in_segment_t &tag, const result_t &res) const
{
	for(size_t i = 0; i < size; ++i) {
		items[i]->commit(request, tag, res);
	}
}

namespace method_elliptics {
config_binding_sname(method_elliptics_t);
config_binding_value(method_elliptics_t, port);
config_binding_value(method_elliptics_t, family);
config_binding_type(method_elliptics_t, elliptics_source_t);
config_binding_value(method_elliptics_t, source);
config_binding_value(method_elliptics_t, logger_filename);
config_binding_value(method_elliptics_t, logger_level);
config_binding_type(method_elliptics_t, logger_t);
config_binding_value(method_elliptics_t, loggers);
config_binding_value(method_elliptics_t, timeout);
config_binding_value(method_elliptics_t, nodes_count);
config_binding_value(method_elliptics_t, flags);
config_binding_value(method_elliptics_t, check_timeout);
config_binding_value(method_elliptics_t, io_thread_num);
config_binding_value(method_elliptics_t, net_thread_num);
//config_binding_ctor(method_t, method_elliptics_t);

namespace ipv4 {
config_binding_sname(method_elliptics_ipv4_t);
config_binding_value(method_elliptics_ipv4_t, address);
config_binding_parent(method_elliptics_ipv4_t, method_elliptics_t);
config_binding_ctor(method_t, method_elliptics_ipv4_t);
}

namespace ipv6 {
config_binding_sname(method_elliptics_ipv6_t);
config_binding_value(method_elliptics_ipv6_t, address);
config_binding_parent(method_elliptics_ipv6_t, method_elliptics_t);
config_binding_ctor(method_t, method_elliptics_ipv6_t);
}

typedef stat::count_t conns_t;
typedef stat::count_t icount_t;
typedef stat::count_t ocount_t;
typedef stat::mmcount_t mmtasks_t;

class load_t {
	spinlock_t spinlock;
	interval_t real, event;

public:
	inline load_t() throw() :
	spinlock(), real(interval::zero), event(interval::zero) { }

	inline ~load_t() throw() { }

	load_t(load_t const &) = delete;
	load_t &operator=(load_t const &) = delete;

	inline void put(interval_t _real, interval_t _event) {
		spinlock_guard_t guard(spinlock);
		real += _real;
		event += _event;
	}

	typedef load_t val_t;

	class res_t {
	public:
		interval_t real, event;

		inline res_t(load_t &load) throw() {
			spinlock_guard_t guard(load.spinlock);
			real = load.real; load.real = interval::zero;
			event = load.event; load.event = interval::zero;
		}

		inline res_t() throw() :
		real(interval::zero), event(interval::zero) { }

		inline ~res_t() throw() { }

		inline res_t(res_t const &) = default;
		inline res_t &operator=(res_t const &) = default;

		inline res_t &operator+=(res_t const &res) throw() {
			real += res.real;
			event += res.event;

			return *this;
		}
	};

	friend class res_t;
};


typedef stat::items_t<
	conns_t,
	icount_t,
	ocount_t,
	mmtasks_t,
	load_t
> stat_base_t;

struct stat_t : stat_base_t {
	inline stat_t() throw() : stat_base_t(
		STRING("conns"),
		STRING("in"),
		STRING("out"),
		STRING("mmtasks"),
		STRING("load")
	) {}

	inline ~stat_t() throw() {}

	inline conns_t &conns() throw() { return item<0>(); }
	inline icount_t &icount() throw() { return item<1>(); }
	inline ocount_t &ocount() throw() { return item<2>(); }
	inline mmtasks_t &mmtasks() throw() { return item<3>(); }
	inline load_t &load() throw() { return item<4>(); }
};

}

static ioremap::elliptics::logger_base create_logger(const method_elliptics_t::config_t &config) {
	try {
		using ioremap::elliptics::logger_base;
		using ioremap::elliptics::log_level;
		using ioremap::elliptics::file_logger;
		MKCSTR(logger_filename, config.logger_filename);
		MKCSTR(logger_level, config.logger_level);

		if (strcmp(logger_filename, "/dev/null") == 0) {
			static_assert(static_cast<int>(DNET_LOG_ERROR) > static_cast<int>(DNET_LOG_DEBUG), "Fix log_level");

			log_level level = static_cast<log_level>(static_cast<int>(DNET_LOG_ERROR) + 1);
			logger_base logger;
			logger.verbosity(level);
			return logger;
		} else {
			log_level level = file_logger::parse_level(logger_level);
			return file_logger(logger_filename, level);
		}
	} catch (const ioremap::elliptics::error &e) {
		throw exception_sys_t(log::error, e.error_code(), "%s", e.what());
	} catch (const std::exception &e) {
		throw exception_sys_t(log::error, 0, "%s", e.what());
	}
}

static dnet_config create_config(const method_elliptics_t::config_t &config) {
	dnet_config cfg;
	memset(&cfg, 0, sizeof(cfg));
	cfg.wait_timeout = config.timeout;
	cfg.flags = config.flags;
	cfg.check_timeout = config.check_timeout;
	cfg.io_thread_num = config.io_thread_num;
	cfg.net_thread_num = config.net_thread_num;
	return cfg;
}

method_elliptics_t::method_elliptics_t(const string_t &, const config_t &config) :
	method_t(STRING("elliptics")), logger(create_logger(config)), cfg(create_config(config)),
	source(*config.source), stat(*new stat_t), loggers(*new loggers_t(config.loggers))
{

	for (uint i = 0; i < config.nodes_count; ++i)
		nodes.emplace_back(ioremap::elliptics::logger(logger, blackhole::log::attributes_t()), cfg);
}

method_elliptics_t::~method_elliptics_t() throw() {
	delete &stat;
	delete &loggers;
}

static inline std::string make_string(in_segment_t &in)
{
	std::string string;
	string.resize(in.size());
	out_t(&string[0], string.size())(in);
	return string;
}

struct method_elliptics_handler_t
{
	ioremap::elliptics::error_info *exception;
	bq_cond_t *cond;
	timeval_t *time_recv;
	size_t *size_in;

	void operator() (const ioremap::elliptics::callback_result_entry &result)
	{
		*size_in += result.raw_data().size();
	}

	void operator() (const ioremap::elliptics::error_info &error)
	{
		bq_cond_t::handler_t handler(*cond);
		*exception = error;
		*time_recv = timeval::current();
		handler.send();
	}
};

bool method_elliptics_t::test(times_t &times) const
{
	request_t request;
	if (!source.get_request(request))
		return false;

	result_t result;

	elliptics_session_t session(nodes[rand() % nodes.size()]);

	session.set_cflags(request.cflags);
	session.set_ioflags(request.ioflags);
	session.set_groups(request.groups);
	session.set_filter(ioremap::elliptics::filters::all_with_ack);
	session.set_exceptions_policy(ioremap::elliptics::session::no_exceptions);

	result.size_in += request.groups.size() * sizeof(dnet_cmd);
	result.size_out += request.groups.size() * sizeof(dnet_cmd);

	ioremap::elliptics::key id;
	if (request.id) {
		MKCSTR(cid, request.id);
		dnet_id did;
		memset(&did, 0, sizeof(did));
		dnet_parse_numeric_id(cid, did.id);
		id = did;
	} else if (request.command == method_elliptics::exec_request) {
		id = make_string(request.data);
	} else {
		id = make_string(request.filename);
	}

	try {
		ioremap::elliptics::error_info exception;
		bq_cond_t cond;
		method_elliptics_handler_t handler = {
			&exception, &cond, &result.time_recv, &result.size_in
		};

		switch (request.command) {
		case method_elliptics::write_data:
			result.size_out += request.groups.size() * request.data.size();
			session.write_data(id, make_string(request.data), request.offset).connect(handler, handler);
			break;
		case method_elliptics::read_data:
			session.read_data(id, request.offset, request.size).connect(handler, handler);
			break;
		case method_elliptics::remove_data:
			session.remove(id).connect(handler, handler);
			break;
		case method_elliptics::exec_request: {
			result.size_out += request.data.size();
			const std::string data = make_string(request.data);
			id.transform(session);
			dnet_id did = id.id();
			session.exec(&did, make_string(request.filename), data).connect(handler, handler);
			break;
			}
		}

		result.time_conn = timeval::current();
		result.time_send = result.time_conn;

		{
			bq_cond_t::handler_t handler(cond);
			if (!result.time_recv.is_real() && !bq_success(handler.wait(NULL)))
				throw exception_sys_t(log::error, errno, "cond.wait: %m");
		}

		result.time_recv = std::max(result.time_recv, result.time_send);
		result.time_end = timeval::current();

		if (exception) {
			result.log_level = logger_t::proto_warning;
			result.err = exception.code();
			switch (exception.code()) {
			case -ENOENT:
				result.proto_code = 404;
				break;
			case -ENXIO:
				result.proto_code = 403;
				break;
			default:
				result.proto_code = 500;
				break;
			}
		} else {
			result.proto_code = 200;
		}

		auto interval_real = result.time_end - result.time_start;

		times.inc(interval_real);
		stat.load().put(interval_real, result.time_end - result.time_send);
		stat.icount() += result.size_in;
		stat.ocount() += result.size_out;

		loggers.commit(request.request, request.tag, result);
	} catch (const ioremap::elliptics::error &e) {
		throw exception_sys_t(log::error, e.error_code(), "%s", e.what());
	} catch (const std::exception &e) {
		throw exception_sys_t(log::error, 0, "%s", e.what());
	}

	return true;
}

void method_elliptics_t::do_init()
{
	source.init();
}

void method_elliptics_t::do_fini()
{
	source.fini();
}

void method_elliptics_t::do_run() const
{
	source.run();
}

void method_elliptics_t::do_stat_print() const
{
	stat.print();
}

method_elliptics_ipv4_t::config_t::config_t() throw()
{
	family = AF_INET;
}

void method_elliptics_ipv4_t::config_t::check(const in_t::ptr_t &ptr) const
{
	if (!address)
		config::error(ptr, "address is required");

	method_elliptics_t::config_t::check(ptr);
}

method_elliptics_ipv4_t::method_elliptics_ipv4_t(const string_t &s, const config_t &c) :
	method_elliptics_t(s, c)
{
	try {
		string_t address_str(string_t::ctor_t((3 + 1) * 4 - 1).print(c.address));
		MKCSTR(address, address_str);
		for (uint i = 0; i < nodes.size(); ++i)
			nodes[i].add_remote(ioremap::elliptics::address(address, c.port, c.family));
	} catch (const ioremap::elliptics::error &e) {
		throw exception_sys_t(log::error, e.error_code(), "%s", e.what());
	} catch (const std::exception &e) {
		throw exception_sys_t(log::error, 0, "%s", e.what());
	}
}

method_elliptics_ipv4_t::~method_elliptics_ipv4_t() throw()
{
}

method_elliptics_ipv6_t::config_t::config_t() throw()
{
	family = AF_INET;
}

void method_elliptics_ipv6_t::config_t::check(const in_t::ptr_t &ptr) const
{
	if (!address)
		config::error(ptr, "address is required");

	method_elliptics_t::config_t::check(ptr);
}

method_elliptics_ipv6_t::method_elliptics_ipv6_t(const string_t &s, const config_t &c) :
	method_elliptics_t(s, c)
{
	try {
		string_t address_str(string_t::ctor_t((4 + 1) * 8 - 1).print(c.address));
		MKCSTR(address, address_str);
		for (uint i = 0; i < nodes.size(); ++i)
			nodes[i].add_remote(ioremap::elliptics::address(address, c.port, c.family));
	} catch (const ioremap::elliptics::error &e) {
		throw exception_sys_t(log::error, e.error_code(), "%s", e.what());
	} catch (const std::exception &e) {
		throw exception_sys_t(log::error, 0, "%s", e.what());
	}
}

method_elliptics_ipv6_t::~method_elliptics_ipv6_t() throw()
{
}

} // namespace io_benchmark
} // namespace phantom


src/method_elliptics.H
// This file is part of the phantom::io_benchmark module.
// Copyright (C) 2013, Ruslan Nigmatullin <euroelessar@yandex.ru>.
// Copyright (C) 2013, YANDEX LLC.
// This module may be distributed under the terms of the GNU LGPL 2.1.
// See the file ‘COPYING’ or ‘http://www.gnu.org/licenses/lgpl-2.1.html’.

#pragma once

#include <phantom/io_benchmark/method.H>

#include <pd/base/fd.H>
#include <pd/base/netaddr.H>
#include <pd/base/config.H>
#include <pd/base/config_list.H>
#include <pd/base/ipv4.H>
#include <pd/base/ipv6.H>

#include <elliptics/cppdef.h>

#pragma GCC visibility push(default)

namespace phantom { namespace io_benchmark {

namespace method_elliptics {
class elliptics_source_t;
class request_t;
class stat_t;
}
namespace method_stream {
class logger_t;
class result_t;
}

class method_elliptics_t : public method_t
{
public:
	typedef ioremap::elliptics::node elliptics_node_t;
	typedef ioremap::elliptics::session elliptics_session_t;
	typedef ioremap::elliptics::file_logger elliptics_file_logger_t;
	typedef ioremap::elliptics::logger_base elliptics_logger_base_t;
	typedef ioremap::elliptics::logger elliptics_logger_t;
	typedef method_elliptics::elliptics_source_t elliptics_source_t;
	typedef method_elliptics::request_t request_t;
	typedef method_elliptics::stat_t stat_t;
	typedef method_stream::logger_t logger_t;
	typedef method_stream::result_t result_t;

protected:
	struct loggers_t : sarray1_t<logger_t *>
	{
		inline loggers_t(const config::list_t<config::objptr_t<logger_t>> &list) throw() : sarray1_t<logger_t *>(list) {}
		inline ~loggers_t() throw() {}

		void commit(const in_segment_t &request, const in_segment_t &tag, const result_t &res) const;

	};

	elliptics_logger_base_t logger;
	dnet_config cfg;
	std::vector<elliptics_node_t> nodes;
	elliptics_source_t &source;
	stat_t &stat;
	loggers_t &loggers;

	bool test(times_t &stat) const;
	void do_init();
	void do_fini();
	void do_run() const;
	void do_stat_print() const;

public:
	struct config_t {
		config_binding_type_ref(elliptics_source_t);
		config_binding_type_ref(logger_t);

		u_int16_t port;
		int family;
		string_t logger_filename;
		string_t logger_level;
		config::objptr_t<elliptics_source_t> source;
		config::list_t<config::objptr_t<logger_t> > loggers;
		uint timeout;
		uint nodes_count;
		long check_timeout;
		int flags;
		int io_thread_num;
		int net_thread_num;

		config_t() throw();
		void check(in_t::ptr_t const &ptr) const;
		inline ~config_t() throw() { }
	};

	method_elliptics_t(string_t const &, config_t const &);
	~method_elliptics_t() throw();
};

class method_elliptics_ipv4_t : public method_elliptics_t
{
public:
	struct config_t : method_elliptics_t::config_t {
		address_ipv4_t address;

		config_t() throw();
		void check(in_t::ptr_t const &ptr) const;
	};

	method_elliptics_ipv4_t(string_t const &, config_t const &);
	~method_elliptics_ipv4_t() throw();
};

class method_elliptics_ipv6_t : public method_elliptics_t
{
public:
	struct config_t : method_elliptics_t::config_t {
		address_ipv6_t address;

		config_t() throw();
		void check(in_t::ptr_t const &ptr) const;
	};

	method_elliptics_ipv6_t(string_t const &, config_t const &);
	~method_elliptics_ipv6_t() throw();
};

}} // namespace phantom::io_benchmark

#pragma GCC visibility pop


src/source_log/source_log.C
// This file is part of the phantom::io_benchmark module.
// Copyright (C) 2006-2012, Eugene Mamchits <mamchits@yandex-team.ru>.
// Copyright (C) 2013, Ruslan Nigmatullin <euroelessar@yandex.ru>.
// Copyright (C) 2006-2013, YANDEX LLC.
// This module may be distributed under the terms of the GNU LGPL 2.1.
// See the file ‘COPYING’ or ‘http://www.gnu.org/licenses/lgpl-2.1.html’.

#include "../source.H"

#include <phantom/module.H>

#include <pd/base/config.H>
#include <pd/base/exception.H>
#include <pd/base/in_fd.H>
#include <pd/base/thr.H>
#include <pd/base/size.H>
#include <pd/base/config.H>
#include <pd/base/cmp.H>
#include <pd/base/mutex.H>

#include <pd/bq/bq_util.H>

#include <unistd.h>
#include <fcntl.h>

#include <iostream>

namespace phantom { namespace io_benchmark { namespace method_elliptics {

enum {
	LINE_LENGTH = 1024,
	TAG_LEN = 256
};

MODULE(io_benchmark_method_elliptics_source_log);

class log_file_t {
	in_fd_t in;
	in_t::ptr_t ptr;
	timeval_t timeval_start;
	interval_t delta, delta_max;
	bool work;

public:
	inline log_file_t(size_t ibuf_size, int fd) :
		in(ibuf_size, fd), ptr(in), timeval_start(timeval::current()),
		delta(interval::zero), delta_max(interval::zero), work(true) { }

	inline ~log_file_t() throw() { }

	bool get_request(request_t &request, interval_t &interval_sleep);

	interval_t stat(bool clear);
};

static void error_handler(in_t::ptr_t const &, char const *msg) {
	throw exception_log_t(log::error, "format error: %s", msg);
}

bool log_file_t::get_request(
	request_t &request, interval_t &interval_sleep
) {
	if (!work)
		return false;

	try {
		size_t data_length = 0;
		ptr.parse(data_length, &error_handler);

		if(!data_length) {
			work = false;
			return false;
		}

		if(*ptr == ' ') {
			++ptr;
			int64_t msec = 0;
			ptr.parse(msec, &error_handler);
			interval_t interval_request = msec * interval::millisecond;

			interval_sleep = interval_request - (timeval::current() - timeval_start);

			if(interval_sleep < interval::zero) {
				if (delta < -interval_sleep) {
					delta = -interval_sleep;

					if(delta_max < delta)
						delta_max = delta;
				}

				interval_sleep = interval::zero;
			}
		}

		if(*ptr == ' ') {
			in_t::ptr_t tagp = ++ptr;
			size_t limit = TAG_LEN;
			if(!ptr.scan("\n", 1, limit))
				throw exception_log_t(log::error, "format error #1");

			limit = ptr - tagp;
			if(tagp.scan("\t", 1, limit))
				throw exception_log_t(log::error, "format error #2");

			request.tag = in_segment_t(tagp, ptr - tagp);
		}
		else {
			request.tag = in_segment_t();
		}

		if(*ptr != '\n')
			throw exception_log_t(log::error, "format error #3");
		++ptr;

		in_t::ptr_t endp = ptr + data_length;

		request.request = in_segment_t(ptr, data_length);

		ptr.parse(request.cflags, &error_handler);
		if(*ptr != ' ')
			throw exception_log_t(log::error, "format error #4");
		++ptr;

		ptr.parse(request.ioflags, &error_handler);
		if(*ptr != ' ')
			throw exception_log_t(log::error, "format error #5");

		request.groups.resize(0);
		int group_id;
		do {
			++ptr;

			ptr.parse(group_id, error_handler);
			request.groups.push_back(group_id);
		} while (*ptr == ':');
		if(*ptr != ' ')
			throw exception_log_t(log::error, "format error #6");
		++ptr;

		in_t::ptr_t filenamep = ptr;
		size_t limit = LINE_LENGTH;
		if (!ptr.scan(" \n", 2, limit))
			throw exception_log_t(log::error, "format error #7");

		request.filename = in_segment_t(filenamep, ptr - filenamep);

		if (*ptr == ' ') {
			++ptr;

			in_t::ptr_t idp = ptr;
			limit = LINE_LENGTH;
			if (!ptr.scan(" \n", 2, limit))
				throw exception_log_t(log::error, "format error #7-2");

			request.id = in_segment_t(idp, ptr - idp);
		}

		if (*ptr != '\n')
			throw exception_log_t(log::error, "format error #8");
		++ptr;

		limit = LINE_LENGTH;
		in_t::ptr_t linep = ptr;
		ptr.scan(" \n", 2, limit);

		str_t write_command = CSTR("write");
		str_t read_command = CSTR("read");
		str_t remove_command = CSTR("remove");
		str_t exec_command = CSTR("exec");

		MKCSTR(cmd, in_segment_t(linep, 4));

		if (linep.match<ident_t>(write_command)) {
			linep += write_command.size();
			request.command = write_data;
		} else if (linep.match<ident_t>(read_command)) {
			linep += read_command.size();
			request.command = read_data;
		} else if (linep.match<ident_t>(remove_command)) {
			linep += remove_command.size();
			request.command = remove_data;
		} else if (linep.match<ident_t>(exec_command)) {
			linep += exec_command.size();
			request.command = exec_request;
		} else {
			throw exception_log_t(log::error, "format error #9");
		}
		if (*ptr != ' ' && *ptr != '\n')
			throw exception_log_t(log::error, "format error #10");

		if (*ptr == ' ') {
			++ptr;
			ptr.parse(request.offset, error_handler);
		} else {
			request.offset = 0;
		}

		if (*ptr == ' ') {
			++ptr;
			ptr.parse(request.size, error_handler);
		} else {
			request.size = 0;
		}

		if (*ptr != '\n')
			throw exception_log_t(log::error, "format error #11");
		++ptr;

		if (request.command == write_data || request.command == exec_request) {
			if (ptr == endp)
				throw exception_log_t(log::error, "format error #12");

			request.data = in_segment_t(ptr, endp - ptr - 1);
		} else if (ptr != endp) {
			throw exception_log_t(log::error, "format error #13");
		}

		ptr = endp;

		if (*ptr != '\n')
			throw exception_log_t(log::error, "format error #14");

		++ptr;
		if (*ptr == '\n')
			throw exception_log_t(log::error, "format error #15");

		in.truncate(ptr);
	}
	catch(const exception_t &exc) {
		work = false;
		throw;
	}

	return true;
}

interval_t log_file_t::stat(bool clear) {
	interval_t delta_print;

	if(clear) {
		delta_print = delta;
		delta = interval::zero;
	}
	else
		delta_print = delta_max;

	return delta_print;
}

class elliptics_source_log_t : public elliptics_source_t {
public:
	struct config_t {
		string_t filename;
		sizeval_t ibuf_size;

		inline config_t() throw() : filename(), ibuf_size(sizeval::mega) { }

		inline void check(in_t::ptr_t const &ptr) const {
			if(!filename)
				config::error(ptr, "filename is required");

			if(ibuf_size > sizeval::giga)
				config::error(ptr, "ibuf_size is too big");

			if(ibuf_size < sizeval::kilo)
				config::error(ptr, "ibuf_size is too small");
		}
	};

private:
	mutable mutex_t mutex;
	int fd;
	size_t ibuf_size;
	string_t filename;
	log_file_t *log_file;


	virtual bool get_request(request_t &request) const;
	virtual void init();
	virtual void run();
	virtual void stat_print() const;
	virtual void fini();
	typedef stat::mminterval_t delta_t;

	typedef stat::items_t<delta_t> stat_base_t;

	struct stat_t : stat_base_t {
		inline stat_t() throw() : stat_base_t(
		STRING("delta")
		) { }

		inline ~stat_t() throw() { }

		inline delta_t &delta() throw() { return item<0>(); }
	};

	stat_t mutable stat;


public:
	inline elliptics_source_log_t(string_t const &, config_t const &config) :
		mutex(), fd(-1), ibuf_size(config.ibuf_size), filename(config.filename),
		log_file(NULL) { }

	inline ~elliptics_source_log_t() throw() { }
};

namespace elliptics_source_log {
config_binding_sname(elliptics_source_log_t);
config_binding_value(elliptics_source_log_t, filename);
config_binding_value(elliptics_source_log_t, ibuf_size);
config_binding_ctor(elliptics_source_t, elliptics_source_log_t);
}

void elliptics_source_log_t::init() {
	MKCSTR(_filename, filename);

	fd = open(_filename, O_RDONLY, 0);
	if(fd < 0)
		throw exception_sys_t(log::error, errno, "open (%s): %m", _filename);

	log_file = new log_file_t(ibuf_size, fd);
}

void elliptics_source_log_t::run()
{
}

void elliptics_source_log_t::stat_print() const
{
	stat.print();
}

bool elliptics_source_log_t::get_request(request_t &request) const {
	interval_t interval_sleep = interval::zero;
	bool res = ({
		mutex_guard_t guard(mutex);

		if(!log_file)
			return false;

		log_file->get_request(request, interval_sleep);
	});

	if (res) {
		if (interval_sleep > interval::zero) {
			stat.delta() = interval::zero;
			if (bq_sleep(&interval_sleep) < 0)
				return false;
		} else {
			stat.delta() = -interval_sleep;
		}
	}

	return res;
}

void elliptics_source_log_t::fini() {
	log_file_t *_log_file = NULL;
	{
		mutex_guard_t guard(mutex);
		_log_file = log_file;
		log_file = NULL;
	}

	delete _log_file;
	close(fd);
	fd = -1;
}

}}} // namespace phantom::io_benchmark::method_elliptics

